src/dragonball/src/api/v1/vmm_action.rs

```diff
@@ -634,3 +634,626 @@ fn handle_cpu_topology(
+    #[test]
+    fn test_vmm_action_insert_network_device() {
+        skip_if_not_root!();
+
+        // hotplug unready
+        let req = VmmAction::InsertNetworkDevice(VirtioNetDeviceConfigInfo::default());
+        check_request(req, InstanceState::Running, |result| {
+            if let Err(VmmActionError::StartMicroVm(StartMicroVmError::UpcallMissVsock)) = result {
+                let err_string = format!("{}", result.unwrap_err());
+                let expected_err = String::from(
+                    "failed to boot the VM: \
+                        the upcall client needs a virtio-vsock device for communication",
+                );
+                assert_eq!(err_string, expected_err);
+            } else {
+                panic!();
+            }
+        });
+
+        // success
+        let req = VmmAction::InsertNetworkDevice(VirtioNetDeviceConfigInfo::default());
+        check_request(req, InstanceState::Uninitialized, |result| {
+            assert!(result.is_ok());
+        });
+    }
+
+    #[cfg(feature = "virtio-net")]
+    #[test]
+    fn test_vmm_action_update_network_interface() {
+        skip_if_not_root!();
+
+        // invalid id
+        let update_config = VirtioNetDeviceConfigUpdateInfo {
+            iface_id: String::from("1"),
+            rx_rate_limiter: None,
+            tx_rate_limiter: None,
+        };
+        let req = VmmAction::UpdateNetworkInterface(update_config);
+        check_request(req, InstanceState::Running, |result| {
+            if let Err(VmmActionError::VirtioNet(VirtioNetDeviceError::InvalidIfaceId(_))) = result
+            {
+                let err_string = format!("{}", result.unwrap_err());
+                let expected_err = String::from(
+                    "virtio-net device error: \
+                    invalid virtio-net iface id '1'",
+                );
+                assert_eq!(err_string, expected_err);
+            } else {
+                panic!();
+            }
+        });
+    }
+
+    #[cfg(feature = "virtio-vsock")]
+    #[test]
+    fn test_vmm_action_insert_vsock_device() {
+        skip_if_not_root!();
+
+        // invalid state
+        let req = VmmAction::InsertVsockDevice(VsockDeviceConfigInfo::default());
+        check_request(req, InstanceState::Running, |result| {
+            if let Err(VmmActionError::Vsock(VsockDeviceError::UpdateNotAllowedPostBoot)) = result {
+                let err_string = format!("{}", result.unwrap_err());
+                let expected_err = String::from(
+                    "failed to add virtio-vsock device: \
+                    update operation is not allowed after boot",
+                );
+                assert_eq!(err_string, expected_err);
+            } else {
+                panic!();
+            }
+        });
+
+        // invalid guest_cid
+        let req = VmmAction::InsertVsockDevice(VsockDeviceConfigInfo::default());
+        check_request(req, InstanceState::Uninitialized, |result| {
```

Comment:
All calling of `check_request` have the same pattern, so it is suitable for [table driven test](https://github.com/kata-containers/kata-containers/blob/f30fe86dc1f53b5e964e6f54de28a0ce0a818786/docs/Unit-Test-Advice.md#rust-table-driven-tests), suggestion fix this in this PR or another separated PR.

Reply:
fixed